README.md
# cloudwatch-metrics-client
Python CloudWatch metrics Client library utilising asyncio (along with sync variant)

This is yet another Python CloudWatch client. While there are alternatives, they are
either not supporting asyncio or not actively maintained.

https://github.com/awslabs/cloudwatch-fluent-metrics

https://github.com/peterdemin/awsme

The client collects whatever metrics you feed it with and reports to CloudWatch regularly, though
you can use it for ad-hoc reporting as well

NOTE: aioboto3 is not in the requirements, in case you only need sync version. If running under Python 3.6, you'll
need async_generator>=1.10 as well

Usage:

```python
import asyncio
from cloudwatch_metrics_client.aiocloudwatch import CloudWatchAsyncMetrics, CloudWatchAsyncMetricReporter

async def setup():
  
  reporter = CloudWatchAsyncMetricReporter(report_interval=REPORT INTERVAL)        # seconds
  CloudWatchAsyncMetrics.with_namespace('<YOUR NAMESPACE').with_reporter(reporter)
  # .with_client(aioboto3_cloudwatch_client) is you're not happy with default client
  await reporter.run()
  
  
@CloudWatchAsyncMetrics.monitored_task
async def process_request(request):
  
  # processing logic...
  #
  # request_type = ... 
  # request_specific_data = 
  
  # Set details on monitored call
  CloudWatchAsyncMetrics\
    .with_monitored_dimension('TypeOfRequest', request_type)\
    .with_monitored_dimension('SpecificAttribute', request_specific_data)
  
  # Put particular metric
  await CloudWatchAsyncMetrics.put_metric(
        MetricName='particular-metric-name',
        Unit='Count',
        Dimensions={'Kind': 'metric-kind', 'Sort': 'metric-sort'},
        Value=value
  )
  
  # Record statistical value
  await CloudWatchAsyncMetrics.put_statistic(
        MetricName='particular-metric-name',
        Unit='Count',
        Dimensions={'Kind': 'metric-kind', 'Sort': 'metric-sort'},
        Value=value
   
  # Elapsed time will be recorded  
```

Elapsed time of each call of `process_request` coroutine will be recorded in internal data structure and aggregated 
statistics will be regularly sent to CloudWatch.
Individual metrics could be collected with `put_metric` call - they too would be stored for a while and then sent
to CloudWatch. `put_metric` keeps each unique value and sends list of them along with number of occurrencies of each
value, `put_statistics` only sends aggregated data over a bunch of metrics - Sum, SampleCount, Min and Max
(See https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudwatch.html#CloudWatch.Client.put_metric_data for details)

Reporter has async `flush` method for flushing metric data still not sent to CloudWatch. If regular reporting is not 
needed, just don't call `run` coro and instead call `flush` when it is time to report metrics.

Sync version works in the very same way, it utilises separate thread for reporting.

```python

from cloudwatch_metrics_client.cloudwatch import CloudWatchSyncMetrics, CloudWatchSyncMetricReporter

def setup():

  reporter = CloudWatchSyncMetricReporter(report_interval=REPORT INTERVAL)        # seconds
  CloudWatchSyncMetrics.with_namespace('<YOUR NAMESPACE').with_reporter(reporter)
  reporter.run()

@CloudWatchSyncMetrics.monitored_task  
def process_request(request):

# Same here, but with CloudWatchSync*


```  

Changelog.md
## 0.0.6 (2019-06-20)

- Add debug level setting
- Fix Namespace parameter

## 0.0.5 (2019-06-17)

- Import aioboto3 lazily

## 0.0.4 (2019-06-17)

- Add support for Python 3.6 (of sorts)

## 0.0.3 (2019-06-17)

- Fix setting client

## 0.0.2 (2019-06-17)

- Fix packaging

## 0.0.1 (2019-06-17)

- Initial release


tests/test_cloudwatch.py
import datetime
import time

import boto3
from unittest import TestCase
from mock import MagicMock

from src.cloudwatch_metrics_client.cloudwatch import CloudWatchSyncMetrics, CloudWatchSyncMetricReporter


class TestCloudwatch(TestCase):

    def setUp(self) -> None:

        boto3.client = MagicMock()

        self.reporter = CloudWatchSyncMetricReporter(report_interval=None)

        CloudWatchSyncMetrics.with_namespace('test_namespace').with_reporter(self.reporter)
        CloudWatchSyncMetrics.setup_client()

    def test_recording_metric(self):

        CloudWatchSyncMetrics.put_metric(MetricName='test_metric', Value=100)

        metrics = list(self.reporter.metrics.values())[0].to_repr()

        self.assertEqual('test_metric', metrics['MetricName'])
        self.assertEqual(100, metrics['Values'][0])
        self.assertEqual(1, metrics['Counts'][0])
        self.assertIsNone(metrics.get('Dimensions'))
        self.assertIsInstance(metrics['Timestamp'], datetime.datetime)

    def test_sync_decorator(self):

        @CloudWatchSyncMetrics.monitored_task
        def task():
            time.sleep(0.1)

        task()
        metrics = self.reporter.statistics
        repr = metrics['transaction?'].to_repr()
        self.assertEqual('transaction', repr['MetricName'])
        self.assertIsNone(repr.get('Dimensions'))
        self.assertEqual(1, repr['StatisticValues']['SampleCount'])
        self.assertLess(100000, repr['StatisticValues']['Sum'])
        self.assertGreater(150000, repr['StatisticValues']['Sum'])
        self.assertEqual('Microseconds', repr['Unit'])

    def test_sync_decorator_with_dimensions(self):

        def test():
            task()
            metrics = list(self.reporter.statistics.values())[0].to_repr()
            self.assertEqual('transaction', metrics['MetricName'])
            self.assertDictEqual({'Name': 'Test_Dimension', 'Value': 'ValueX'}, metrics['Dimensions'][0])
            self.assertEqual(1, metrics['StatisticValues']['SampleCount'])
            self.assertLess(100000, metrics['StatisticValues']['Sum'])
            self.assertGreater(150000, metrics['StatisticValues']['Sum'])
            self.assertEqual('Microseconds', metrics['Unit'])

            task2()
            metrics = list(self.reporter.statistics.values())[1].to_repr()
            self.assertEqual('transaction', metrics['MetricName'])
            self.assertDictEqual({'Name': 'Test_Dimension', 'Value': 'ValueY'}, metrics['Dimensions'][0])
            self.assertEqual(1, metrics['StatisticValues']['SampleCount'])
            self.assertLess(100000, metrics['StatisticValues']['Sum'])
            self.assertGreater(150000, metrics['StatisticValues']['Sum'])
            self.assertEqual('Microseconds', metrics['Unit'])

        @CloudWatchSyncMetrics.monitored_task
        def task():
            CloudWatchSyncMetrics.with_monitored_dimension('Test_Dimension', 'ValueX')
            time.sleep(0.1)

        @CloudWatchSyncMetrics.monitored_task
        def task2():
            CloudWatchSyncMetrics.with_monitored_dimension('Test_Dimension', 'ValueY')
            time.sleep(0.1)

        test()
        metrics = list(self.reporter.statistics.values())
        self.assertEqual(2, len(metrics))

    def test_reporter(self):

        self.stored_kwargs = None

        def put_data(**kwargs):
            self.stored_kwargs = kwargs
            return {'ResponseMetadata': {'HTTPStatusCode': 200}}

        CloudWatchSyncMetrics.client = MagicMock()
        CloudWatchSyncMetrics.client.put_metric_data = put_data

        def test():
            reporter = CloudWatchSyncMetricReporter(report_interval=0.5)
            CloudWatchSyncMetrics.with_reporter(reporter)
            reporter.run()
            CloudWatchSyncMetrics.put_metric(MetricName='test_metric_2', Value=50)
            self.assertIsNone(self.stored_kwargs)
            time.sleep(0.7)
            reporter.stop()
            self.assertEqual('test_metric_2', self.stored_kwargs['MetricData'][0]['MetricName'])
            self.assertEqual(50, self.stored_kwargs['MetricData'][0]['Values'][0])
            self.assertEqual(1, self.stored_kwargs['MetricData'][0]['Counts'][0])

        test()


tests/test_aiocloudwatch.py
import asyncio
import datetime
from time import sleep

import aioboto3
from unittest import TestCase
from mock import MagicMock

from src.cloudwatch_metrics_client.aiocloudwatch import CloudWatchAsyncMetrics, CloudWatchAsyncMetricReporter, MetricDimension, Metric, \
    MetricSeries, StatisticSeries


class TestAsyncCloudWatchReporter(TestCase):

    def setUp(self) -> None:
        self.dimensions = {'Dimension1': 'Test1', 'Dimension2': 'Test2'}

    def test_metric_dimension(self):

        dim = MetricDimension({'Dimension': 'Test'})
        self.assertListEqual([{'Name': 'Dimension', 'Value':'Test'}], dim.to_repr())
        self.assertEqual('Dimension=Test', dim.dim_id)

        dim = MetricDimension(dimensions=None)
        self.assertIsNone(dim.to_repr())
        self.assertEqual('', dim.dim_id)

        dim = MetricDimension(self.dimensions)
        self.assertListEqual([{'Name': 'Dimension1', 'Value': 'Test1'}, {'Name': 'Dimension2', 'Value': 'Test2'}],
                             dim.to_repr())
        self.assertEqual('Dimension1=Test1&Dimension2=Test2', dim.dim_id)

    def test_metric(self):

        metric = Metric(name='Metric0', dimensions=self.dimensions, value=15, unit='furlong')
        repr = metric.to_repr()
        self.assertEqual('Metric0', repr['MetricName'])
        self.assertEqual(15, repr['Value'])
        self.assertEqual(2, len(repr['Dimensions']))
        self.assertEqual('furlong', repr['Unit'])
        self.assertEqual('Metric0?Dimension1=Test1&Dimension2=Test2', metric.metric_id)

    def test_metric_series(self):

        metric_series = MetricSeries(name='Series0', dimensions=self.dimensions)
        for n in range(5):
            metric_series.add_value(n)
        metric_series.add_value(2)
        metric_series.add_value(10)
        metric_series.add_value(2)
        repr = metric_series.to_repr()
        self.assertEqual('Series0',repr['MetricName'])
        self.assertEqual(2, len(repr['Dimensions']))
        self.assertListEqual([0, 1, 2, 3, 4, 10], repr['Values'])
        self.assertListEqual([1, 1, 3, 1, 1, 1], repr['Counts'])

    def test_statistic_series(self):

        statistic_series = StatisticSeries(name='Series0', dimensions=self.dimensions)
        for n in range(5):
            statistic_series.add_value(n)
        statistic_series.add_value(2)
        statistic_series.add_value(10)
        statistic_series.add_value(2)
        repr = statistic_series.to_repr()
        self.assertEqual('Series0',repr['MetricName'])
        self.assertEqual(2, len(repr['Dimensions']))
        self.assertEqual(8, repr['StatisticValues']['SampleCount'])
        self.assertEqual(24, repr['StatisticValues']['Sum'])
        self.assertEqual(0, repr['StatisticValues']['Minimum'])
        self.assertEqual(10, repr['StatisticValues']['Maximum'])


class TestCloudwatch(TestCase):

    def setUp(self) -> None:

        aioboto3.client = MagicMock()

        self.reporter = CloudWatchAsyncMetricReporter(report_interval=None)

        CloudWatchAsyncMetrics.with_namespace('test_namespace').with_reporter(self.reporter)
        CloudWatchAsyncMetrics.setup_client()

    def test_recording_metric(self):

        async def test():
            await CloudWatchAsyncMetrics.put_metric(MetricName='test_metric', Value=100)

            metrics = list(self.reporter.metrics.values())[0].to_repr()

            self.assertEqual('test_metric', metrics['MetricName'])
            self.assertEqual(100, metrics['Values'][0])
            self.assertEqual(1, metrics['Counts'][0])
            self.assertIsNone(metrics.get('Dimensions'))
            self.assertIsInstance(metrics['Timestamp'], datetime.datetime)

        asyncio.get_event_loop().run_until_complete(test())

    def test_async_decorator(self):

        async def test():
            await task()

        @CloudWatchAsyncMetrics.monitored_task
        async def task():
            await asyncio.sleep(0.1)

        asyncio.get_event_loop().run_until_complete(test())
        metrics = self.reporter.statistics
        repr = metrics['transaction?'].to_repr()
        self.assertEqual('transaction', repr['MetricName'])
        self.assertIsNone(repr.get('Dimensions'))
        self.assertEqual(1, repr['StatisticValues']['SampleCount'])
        self.assertLess(100000, repr['StatisticValues']['Sum'])
        self.assertGreater(150000, repr['StatisticValues']['Sum'])
        self.assertEqual('Microseconds', repr['Unit'])

    def test_async_decorator_with_dimensions(self):

        async def test():
            await task()
            metrics = list(self.reporter.statistics.values())[0].to_repr()
            self.assertEqual('transaction', metrics['MetricName'])
            self.assertDictEqual({'Name': 'Test_Dimension', 'Value': 'ValueX'}, metrics['Dimensions'][0])
            self.assertEqual(1, metrics['StatisticValues']['SampleCount'])
            self.assertLess(100000, metrics['StatisticValues']['Sum'])
            self.assertGreater(150000, metrics['StatisticValues']['Sum'])
            self.assertEqual('Microseconds', metrics['Unit'])

            await task2()
            metrics = list(self.reporter.statistics.values())[1].to_repr()
            self.assertEqual('transaction', metrics['MetricName'])
            self.assertDictEqual({'Name': 'Test_Dimension', 'Value': 'ValueY'}, metrics['Dimensions'][0])
            self.assertEqual(1, metrics['StatisticValues']['SampleCount'])
            self.assertLess(100000, metrics['StatisticValues']['Sum'])
            self.assertGreater(150000, metrics['StatisticValues']['Sum'])
            self.assertEqual('Microseconds', metrics['Unit'])

        @CloudWatchAsyncMetrics.monitored_task
        async def task():
            CloudWatchAsyncMetrics.with_monitored_dimension('Test_Dimension', 'ValueX')
            await asyncio.sleep(0.1)

        @CloudWatchAsyncMetrics.monitored_task
        async def task2():
            CloudWatchAsyncMetrics.with_monitored_dimension('Test_Dimension', 'ValueY')
            await asyncio.sleep(0.1)

        asyncio.get_event_loop().run_until_complete(test())
        metrics = list(self.reporter.statistics.values())
        self.assertEqual(2, len(metrics))


    def test_sync_decorator_in_coro(self):

        @CloudWatchAsyncMetrics.monitored_task
        def task():
            sleep(0.15)

        async def test():
            task()
            await asyncio.sleep(0.1)

        asyncio.get_event_loop().run_until_complete(test())

        metrics = list(self.reporter.statistics.values())[0].to_repr()
        self.assertEqual('transaction', metrics['MetricName'])
        self.assertIsNone(metrics.get('Dimensions'))
        self.assertEqual(1, metrics['StatisticValues']['SampleCount'])
        self.assertLess(150000, metrics['StatisticValues']['Sum'])
        self.assertGreater(200000, metrics['StatisticValues']['Sum'])

    def test_reporter(self):

        self.stored_kwargs = None

        async def put_data(**kwargs):
            self.stored_kwargs = kwargs
            return {'ResponseMetadata': { 'HTTPStatusCode': 200 }}

        CloudWatchAsyncMetrics.client = MagicMock()
        CloudWatchAsyncMetrics.client.put_metric_data = put_data

        async def test():
            reporter = CloudWatchAsyncMetricReporter(report_interval=0.5)
            CloudWatchAsyncMetrics.with_reporter(reporter)
            await reporter.run()
            await CloudWatchAsyncMetrics.put_metric(MetricName='test_metric_2', Value=50)
            self.assertIsNone(self.stored_kwargs)
            await asyncio.sleep(0.7)
            self.assertEqual('test_metric_2', self.stored_kwargs['MetricData'][0]['MetricName'])
            self.assertEqual(50, self.stored_kwargs['MetricData'][0]['Values'][0])
            self.assertEqual(1, self.stored_kwargs['MetricData'][0]['Counts'][0])

        asyncio.get_event_loop().run_until_complete(test())


src/cloudwatch_metrics_client/cloudwatch.py
import datetime
import functools
import logging
import math
import threading
from contextlib import contextmanager

import boto3

from cloudwatch_metrics_client.aiocloudwatch import CloudWatchAsyncMetrics, Metric, MetricSeries, StatisticSeries

log = logging.getLogger(__name__.split('.')[0])


class CloudWatchSyncMetrics(CloudWatchAsyncMetrics):

    @classmethod
    def setup_client(cls):
        if cls.client is None:
            cls.client = boto3.client('cloudwatch')
        return cls

    @classmethod
    def put_metric(cls, **metric_data):
        try:
            return cls.reporter.put_metric(**metric_data)
        except AttributeError:
            return False

    @classmethod
    def put_statistic(cls, name, dimensions, value, unit=None):
        try:
            return cls.reporter.put_statistic(name, dimensions, value, unit)
        except AttributeError:
            return False

    @classmethod
    def send_metric(cls, **metric_data):

        try:
            if metric_data.get('Timestamp') is None:
                metric_data['Timestamp'] = datetime.datetime.now()
            cls.setup_client()
            return cls.client.put_metric_data(
                Namespace=cls.namespace,
                MetricData=[{**metric_data}]
            )
        except AttributeError:
            return False

    @classmethod
    def monitored_task(cls, func, name='transaction'):

        @contextmanager
        def monitor():
            previous_dimensions = cls.dimensions
            cls.dimensions = None
            start = datetime.datetime.now()
            yield
            elapsed = datetime.datetime.now() - start
            cls.put_statistic(
                    name=name, dimensions=cls.dimensions, value=elapsed.microseconds, unit='Microseconds')
            cls.dimensions = previous_dimensions

        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            with monitor():
                return func(*args, **kwargs)
        return wrapper


class CloudWatchSyncMetricReporter:

    MAX_METRICS_PER_REPORT = 20

    def __init__(self, report_interval=30):

        self.metrics = {}
        self.statistics = {}
        self.report_interval = report_interval
        self.timer = None
        self.report_task = None

        self.lock = threading.Lock()

        self.stopped = False

    def run(self):
        if self.report_interval:
            self.report_task = threading.Thread(target=self.report)
            self.report_task.start()
            log.debug('Reporting metrics to CloudWatch every {} sec'.format(self.report_interval))

    def stop(self):
        if self.timer is not None:
            self.timer.set()
        self.stopped = True

    def put_metric(self, **metric_data):
        with self.lock:
            name = metric_data['MetricName']
            dimensions = metric_data.get('Dimensions')
            metric_id = Metric.generate_id(name, dimensions)
            metric = self.metrics.get(metric_id)
            if metric is None:
                self.metrics[metric_id] = MetricSeries(name=name, dimensions=dimensions, unit=metric_data.get('Unit'))
                metric = self.metrics[metric_id]
            metric.add_value(metric_data['Value'])

        return True

    def put_statistic(self, name, dimensions, value, unit=None):
        with self.lock:
            metric_id = Metric.generate_id(name, dimensions)
            stat = self.statistics.get(metric_id)
            if stat is None:
                self.statistics[metric_id] = StatisticSeries(name=name, dimensions=dimensions, unit=unit)
                stat = self.statistics[metric_id]
            stat.add_value(value)

        return True

    def report(self):
        while True:
            try:
                self.timer = threading.Event()
                self.stopped = self.timer.wait(self.report_interval)
                if self.stopped:
                    log.debug('reporter stopped')
                    return
                if len(self.metrics) + len(self.statistics) == 0:
                    log.debug('nothing to report')
                    continue
                self._report()
            except Exception as e:
                log.error(e)

    def _calculate_statistics(self) -> []:
        statistics = [stat.to_repr() for name, stat in self.statistics.items()]
        self.statistics = {}
        return list(filter(None.__ne__, statistics))

    def _calculate_metrics(self) -> []:
        metrics = [data.to_repr() for name, data in self.metrics.items()]
        self.metrics = {}
        return list(filter(None.__ne__, metrics))

    def _report(self):
        CloudWatchSyncMetrics.setup_client()
        with self.lock:
            num_metrics = len(self.metrics) + len(self.statistics)
            metric_data = self._calculate_metrics() + self._calculate_statistics()
            for n in range(math.ceil(len(metric_data) / CloudWatchSyncMetricReporter.MAX_METRICS_PER_REPORT)):
                if CloudWatchSyncMetrics.debug_level > 1:
                    log.debug('Namespace: {}'.format(CloudWatchSyncMetrics.namespace))
                    log.debug('Metric data: {}'.format(
                        metric_data[n * CloudWatchSyncMetricReporter.MAX_METRICS_PER_REPORT:
                                           (n + 1) * CloudWatchSyncMetricReporter.MAX_METRICS_PER_REPORT]))
                response = CloudWatchSyncMetrics.client.put_metric_data(
                    Namespace=CloudWatchSyncMetrics.namespace,
                    MetricData=metric_data[n * CloudWatchSyncMetricReporter.MAX_METRICS_PER_REPORT:
                                           (n + 1) * CloudWatchSyncMetricReporter.MAX_METRICS_PER_REPORT]
                )
                if response.get('ResponseMetadata', {}).get('HTTPStatusCode') != 200:
                    log.warning('Failed reporting metrics to CloudWatch; response={}'.format(
                        response
                    ))
            log.debug('Reported {} metrics to CloudWatch'.format(num_metrics))

    def flush(self):
        self._report()


src/cloudwatch_metrics_client/aiocloudwatch.py
import asyncio
import datetime
import functools
import logging
import math
from collections import OrderedDict
from typing import Union
from contextlib import contextmanager

# Support for 3.6, for now add dependency manually
try:
    from contextlib import asynccontextmanager
except ImportError:
    from async_generator import asynccontextmanager


log = logging.getLogger(__name__.split('.')[0])
log.setLevel(logging.INFO)

class CloudWatchAsyncMetrics:

    dimensions = None
    client = None
    namespace = None
    reporter = None
    debug_level = 0

    @classmethod
    def with_namespace(cls, namespace):
        cls.namespace = namespace
        return cls

    @classmethod
    def with_client(cls, client):
        cls.client = client
        return cls

    @classmethod
    def setup_client(cls):
        if cls.client is None:
            import aioboto3
            cls.client = aioboto3.client('cloudwatch')
        return cls

    @classmethod
    def with_reporter(cls, reporter):
        cls.reporter = reporter
        return cls

    @classmethod
    def with_debug_level(cls, level):
        cls.debug_level = level
        if level > 0:
            log.setLevel(logging.DEBUG)
        return cls

    @classmethod
    async def put_metric(cls, **metric_data):
        try:
            return await cls.reporter.put_metric(**metric_data)
        except AttributeError:
            return False

    @classmethod
    async def put_statistic(cls, name, dimensions, value, unit=None):
        try:
            return await cls.reporter.put_statistic(name, dimensions, value, unit)
        except AttributeError:
            return False

    @classmethod
    async def send_metric(cls, **metric_data):

        try:
            if metric_data.get('Timestamp') is None:
                metric_data['Timestamp'] = datetime.datetime.now()
            cls.setup_client()
            return await cls.client.put_metric_data(
                Namespace=cls.namespace,
                MetricData=[{**metric_data}]
            )
        except AttributeError:
            return False

    @classmethod
    def with_monitored_dimension(cls, dimension, value):
        if cls.dimensions is None:
            cls.dimensions = {}
        cls.dimensions[dimension] = value
        return cls

    @classmethod
    def monitored_task(cls, func, name='transaction'):

        @contextmanager
        def monitor():
            previous_dimensions = cls.dimensions
            cls.dimensions = None
            start = datetime.datetime.now()
            yield
            elapsed = datetime.datetime.now() - start
            asyncio.run_coroutine_threadsafe(
                cls.put_statistic(
                    name=name, dimensions=cls.dimensions, value=elapsed.microseconds, unit='Microseconds'),
                asyncio.get_event_loop()
            )
            cls.dimensions = previous_dimensions

        @asynccontextmanager
        async def async_monitor():
            previous_dimensions = cls.dimensions
            cls.dimensions = None
            start = datetime.datetime.now()
            yield
            elapsed = datetime.datetime.now() - start
            await cls.put_statistic(
                name=name, dimensions=cls.dimensions, value=elapsed.microseconds, unit='Microseconds')
            cls.dimensions = previous_dimensions

        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            if not asyncio.iscoroutinefunction(func):
                with monitor():
                    return func(*args, **kwargs)
            else:
                async def tmp():
                    async with async_monitor():
                        return await func(*args, **kwargs)

                return tmp()

        return wrapper


class CloudWatchAsyncMetricReporter:

    MAX_METRICS_PER_REPORT = 20

    def __init__(self, report_interval=30):

        self.metrics = {}
        self.statistics = {}
        self.report_interval = report_interval
        self.sleep_task = None
        self.report_task = None

        self.lock = asyncio.Lock()

        self.stopped = False

    async def run(self):
        if self.report_interval:
            self.report_task = asyncio.create_task(self.report())
            log.debug('Reporting metrics to CloudWatch every {} sec'.format(self.report_interval))

    def stop(self):
        if self.sleep_task is not None:
            self.sleep_task.cancel()
        self.stopped = True

    async def put_metric(self, **metric_data):
        async with self.lock:
            name = metric_data['MetricName']
            dimensions = metric_data.get('Dimensions')
            metric_id = Metric.generate_id(name, dimensions)
            metric = self.metrics.get(metric_id)
            if metric is None:
                self.metrics[metric_id] = MetricSeries(name=name, dimensions=dimensions, unit=metric_data.get('Unit'))
                metric = self.metrics[metric_id]
            metric.add_value(metric_data['Value'])

        return True

    async def put_statistic(self, name, dimensions, value, unit=None):
        async with self.lock:
            metric_id = Metric.generate_id(name, dimensions)
            stat = self.statistics.get(metric_id)
            if stat is None:
                self.statistics[metric_id] = StatisticSeries(name=name, dimensions=dimensions, unit=unit)
                stat = self.statistics[metric_id]
            stat.add_value(value)

        return True

    async def report(self):

        while True:
            try:
                if self.stopped:
                    log.debug('reporter stopped')
                    return
                self.sleep_task = asyncio.create_task(asyncio.sleep(self.report_interval))
                try:
                    await self.sleep_task
                except asyncio.CancelledError:
                    log.debug('sleep cancelled; reporter stopped')
                    return
                if len(self.metrics) + len(self.statistics) == 0:
                    log.debug('nothing to report')
                    continue
                await self._report()
            except Exception as e:
                log.error(e)

    def _calculate_statistics(self) -> []:
        statistics = [stat.to_repr() for name, stat in self.statistics.items()]
        self.statistics = {}
        return list(filter(None.__ne__, statistics))

    def _calculate_metrics(self) -> []:
        metrics = [data.to_repr() for name, data in self.metrics.items()]
        self.metrics = {}
        return list(filter(None.__ne__, metrics))

    async def _report(self):
        CloudWatchAsyncMetrics.setup_client()
        async with self.lock:
            num_metrics = len(self.metrics) + len(self.statistics)
            metric_data = self._calculate_metrics() + self._calculate_statistics()
            for n in range(math.ceil(len(metric_data) / CloudWatchAsyncMetricReporter.MAX_METRICS_PER_REPORT)):
                if CloudWatchAsyncMetrics.debug_level > 1:
                    log.debug('Namespace: {})'.format(CloudWatchAsyncMetrics.namespace))
                    log.debug('Metric data: {}'.format(
                        metric_data[n * CloudWatchAsyncMetricReporter.MAX_METRICS_PER_REPORT:
                                           (n + 1) * CloudWatchAsyncMetricReporter.MAX_METRICS_PER_REPORT]))
                response = await CloudWatchAsyncMetrics.client.put_metric_data(
                    Namespace=CloudWatchAsyncMetrics.namespace,
                    MetricData=metric_data[n * CloudWatchAsyncMetricReporter.MAX_METRICS_PER_REPORT:
                                           (n + 1) * CloudWatchAsyncMetricReporter.MAX_METRICS_PER_REPORT]
                )
                if response.get('ResponseMetadata', {}).get('HTTPStatusCode') != 200:
                    log.warning('Failed reporting metrics to CloudWatch; response={}'.format(
                        response
                    ))
            log.debug('Reported {} metrics to CloudWatch'.format(num_metrics))

    async def flush(self):
        await self._report()


class MetricDimension:

    def __init__(self, dimensions):

        self.dimensions = OrderedDict(dimensions) if dimensions is not None else None
        self.dim_id = MetricDimension.generate_id(self.dimensions)

    def to_repr(self) -> Union[list, None]:
        if self.dimensions is None:
            return None
        return [{'Name': name, 'Value': value} for name, value in self.dimensions.items()]

    @staticmethod
    def generate_id(dimensions) -> str:
        if dimensions is not None:
            return '&'.join(['%s=%s' % (a, b) for a, b in dimensions.items()])
        else:
            return ''


class Metric:

    def __init__(self, name, dimensions, value, unit=None):

        self.name = name
        self.dimensions = MetricDimension(dimensions)
        self.value = value
        self.unit = unit

        self.metric_id = Metric.generate_id(self.name, self.dimensions)

    def to_repr(self) -> dict:
        data = {
            'MetricName': self.name,
            'Timestamp': datetime.datetime.now(),
            'Value': self.value
        }
        dimensions = self.dimensions.to_repr()
        if dimensions is not None:
            data['Dimensions'] = dimensions
        if self.unit is not None:
            data['Unit'] = self.unit
        return data

    @staticmethod
    def generate_id(name: str, dimensions) -> str:
        if isinstance(dimensions, MetricDimension):
            return '%s?%s' % (name, dimensions.dim_id)
        else:
            return '%s?%s' % (name, MetricDimension.generate_id(dimensions))


class MetricSeries:

    def __init__(self, name, dimensions=None, unit=None):

        self.metric = Metric(name=name, dimensions=dimensions, value={}, unit=unit)
        self.metric_id = self.metric.metric_id

    def add_value(self, value) -> None:
        values = self.metric.value
        count = values.get(value, 0) + 1
        values[value] = count

    def to_repr(self) -> Union[dict, None]:
        data = self.metric.to_repr()
        del data['Value']
        data['Values'] = []
        data['Counts'] = []
        for value, count in self.metric.value.items():
            data['Values'].append(value)
            data['Counts'].append(count)

        return data


class StatisticSeries(MetricSeries):

    def __init__(self, name, dimensions=None, unit=None):

        super().__init__(name=name, dimensions=dimensions, unit=unit)
        self.metric.value = []

    def to_repr(self) -> Union[dict, None]:
        if len(self.metric.value) == 0:
            return None
        data = self.metric.to_repr()
        del data['Value']
        data['StatisticValues'] = {
            'SampleCount': len(self.metric.value),
            'Sum': sum(self.metric.value),
            'Minimum': min(self.metric.value),
            'Maximum': max(self.metric.value)
        }

        return data

    def add_value(self, value) -> None:
        self.metric.value.append(value)
